ebpf_extensions/neteventebpfext/netevent_ebpf_ext_event.c

```diff
@@ -0,0 +1,525 @@
+
+    if (!netevent_event_context) {
+        goto Exit;
+    }
+
+    if (context_out != NULL && *context_size_out >= sizeof(netevent_event_md_t)) {
+        // Copy the context to the caller.
+        memcpy(netevent_event_context_out, netevent_event_context, sizeof(netevent_event_md_t));
+        *context_size_out = sizeof(netevent_event_md_t);
+
+        // Zero out the event context info.
+        netevent_event_context_out->event_data_start = 0;
+        netevent_event_context_out->event_data_end = 0;
+        *context_size_out = sizeof(netevent_event_md_t);
+    } else {
+        *context_size_out = 0;
+    }
+
+    // Copy the event data to 'data_out'.
+    if (data_out != NULL &&
+        *data_size_out >= (size_t)(netevent_event_context->event_data_end - netevent_event_context->event_data_start)) {
+        memcpy(
+            data_out,
+            netevent_event_context->event_data_start,
+            netevent_event_context->event_data_end - netevent_event_context->event_data_start + 1);
+        *data_size_out = netevent_event_context->event_data_end - netevent_event_context->event_data_start;
+    } else {
+        *data_size_out = 0;
+    }
+
+    ExFreePool(netevent_event_context);
+
+Exit:
+    EBPF_EXT_LOG_EXIT();
+}
+
+//
+// Event Hook NPI client helper functions (invoked by NetEvent as the NPI provider).
+//
+typedef struct _netevent_event_notify_context
+{
+    netevent_event_md_t netevent_event_md;
+} netevent_event_notify_context_t;
+
+void
+_ebpf_netevent_push_event(_In_ netevent_event_md_t* netevent_event)
+{
+    // Logging may delay the event processing, consider enabling only for debugging or if the calling frequency for a
+    // specific use case is low.
+    // EBPF_EXT_LOG_ENTRY();
+
+    if (netevent_event == NULL) {
+        return;
+    }
+
+    ebpf_result_t result;
+    ebpf_extension_hook_client_t* client_context = NULL;
+    netevent_event_notify_context_t netevent_event_notify_context = {0};
+    uint64_t event_size = netevent_event->event_data_end - netevent_event->event_data_start;
+
+    // Currently, the verifier does not support read-only contexts, so we need to copy the event data, rather than
+    // directly passing the existing pointers.
+    // Verifier feature proposal: https://github.com/vbpf/ebpf-verifier/issues/639
+    if (event_size > _event_buffer_size) {
+        // If the event buffer is too small, attempt to resize it.
+        uint8_t* new_event_buffer =
+            (uint8_t*)ExAllocatePoolUninitialized(NonPagedPoolNx, event_size, EBPF_EXTENSION_POOL_TAG);
+        if (new_event_buffer == NULL) {
+            EBPF_EXT_LOG_MESSAGE(
+                EBPF_EXT_TRACELOG_LEVEL_ERROR,
+                EBPF_EXT_TRACELOG_KEYWORD_NETEVENT,
+                "Failed to resize the event buffer - event lost");
+            return;
+        }
+        if (_event_buffer) {
+            ExFreePool(_event_buffer);
+        }
+        _event_buffer = new_event_buffer;
```

Comment:
was curious about why updating the buffer pointer and size was not done under lock?

Reply:
This function is called synchronously from the provider.